pkg/catalog/routes.go

```diff
@@ -514,9 +529,153 @@ func (mc *MeshCatalog) listMeshServices() []service.MeshService {
 	return services
 }
 
+func (mc *MeshCatalog) getDestinationServicesFromTrafficTarget(t *target.TrafficTarget) ([]service.MeshService, error) {
+	sa := service.K8sServiceAccount{
+		Name:      t.Spec.Destination.Name,
+		Namespace: t.Spec.Destination.Namespace,
+	}
+	destServices, err := mc.GetServicesForServiceAccount(sa)
+	if err != nil {
+		return nil, errors.Errorf("Error finding Services for Service Account %#v: %v", sa, err)
+	}
+	return destServices, nil
+}
+
+// getLocalAndNamespacedScopedHostnames returns the local hostnames for the given service and the namespaced scoped hostnames
+//	for the given service
+func (mc *MeshCatalog) getLocalAndNamespacedHostnames(meshService service.MeshService) ([]string, []string, error) {
+	svc := mc.kubeController.GetService(meshService)
+	if svc == nil {
+		return nil, nil, errors.Errorf("Error fetching service %q", meshService)
+	}
+
+	hostnamesSameNamespace := kubernetes.GetLocalNamespaceHostnamesForService(svc)
+	hostnamesDiffNamespace := kubernetes.GetNamespaceScopedHostnamesForService(svc)
+
+	return hostnamesSameNamespace, hostnamesDiffNamespace, nil
+}
+
+func (mc *MeshCatalog) buildInboundPolicies(t *target.TrafficTarget) []*trafficpolicy.InboundTrafficPolicy {
+	inboundPolicies := []*trafficpolicy.InboundTrafficPolicy{}
+
+	// fetch services running workloads with destination service account
+	destServices, err := mc.getDestinationServicesFromTrafficTarget(t)
+	if err != nil {
+		log.Error().Msgf("Error resolving destination from traffic target %s (%s): %v", t.Name, t.Namespace, err)
+		return inboundPolicies
+	}
+
+	// fetch all routes referenced in traffic target
+	routeMatches, err := mc.routesFromRules(t.Spec.Rules, t.Namespace)
+	if err != nil {
+		log.Error().Msgf("Error finding route matches from TrafficTarget %s in namespace %s: %v", t.Name, t.Namespace, err)
+		return inboundPolicies
+	}
+
+	for _, destService := range destServices {
+		wc := getDefaultWeightedClusterForService(destService)
+
+		hostnamesInNS, hostnamesNotInNS, err := mc.getLocalAndNamespacedHostnames(destService)
+		if err != nil {
+			continue
+		}
+
+		// create traffic policy with domains without namespace suffix
+		policyInNS := trafficpolicy.NewInboundTrafficPolicy(buildPolicyName(destService, true), hostnamesInNS)
+		// create traffic policy for domains with namespace suffix
+		policyNotInNS := trafficpolicy.NewInboundTrafficPolicy(buildPolicyName(destService, false), hostnamesNotInNS)
+
+		for _, sourceServiceAccount := range trafficTargetIdentitiesToSvcAccounts(t.Spec.Sources) {
+			for _, routeMatch := range routeMatches {
+				policyNotInNS.AddRule(*trafficpolicy.NewRouteWeightedCluster(routeMatch, wc), sourceServiceAccount)
+				if sourceServiceAccount.Namespace == destService.Namespace {
```

Comment:
I'm assuming this logic right here is why you have the hostnames in two separate lists. That way a local service that is not attached to an authorized Service Account won't accidentally be allowed on the inbound path.

Reply:
yes but check out this [thread](https://github.com/openservicemesh/osm/pull/2187#discussion_r541103978). I went back and forth about it with @snehachhabria and @shashankram and am going to take this out bc of the scenario listed in the example. Lmk what you think.